genmc-sys/src_cpp/MiriInterface.cpp

```diff
@@ -2,49 +2,254 @@
+	auto pos = incPos(parent_id);
+
+	// FIXME(genmc): for supporting symmetry reduction, these will need to be properly set:
+	const unsigned funId = 0;
+	const SVal arg = SVal(0);
+	const ThreadInfo childInfo = ThreadInfo{thread_id, parent_id, funId, arg};
+
+	// NOTE: Default memory ordering (`Release`) used here.
+	auto tcLab = std::make_unique<ThreadCreateLabel>(pos, childInfo);
+	auto createLab = GenMCDriver::handleThreadCreate(std::move(tcLab));
+	auto genmcTid = createLab->getChildId();
+
+	BUG_ON(genmcTid != thread_id || genmcTid == -1 || genmcTid != globalInstructions.size());
+	globalInstructions.push_back(Action(ActionKind::Load, Event(genmcTid, 0)));
+}
+
+void MiriGenMCShim::handleThreadJoin(ThreadId thread_id, ThreadId child_id)
+{
+	auto parentTid = thread_id;
+	auto childTid = child_id;
+
+	// The thread join event happens in the parent.
+	auto pos = incPos(parentTid);
+
+	// NOTE: Default memory ordering (`Acquire`) used here.
+	auto lab = std::make_unique<ThreadJoinLabel>(pos, childTid);
+	auto res = GenMCDriver::handleThreadJoin(std::move(lab));
+	// If the join failed, decrease the event index again:
+	if (!res.has_value())
+		decPos(parentTid);
+
+	// NOTE: Thread return value is ignored, since Miri doesn't need it.
+}
+
+void MiriGenMCShim::handleThreadFinish(ThreadId thread_id, uint64_t ret_val)
+{
+	auto pos = incPos(thread_id);
+	auto retVal = SVal(ret_val);
+
+	// NOTE: Default memory ordering (`Release`) used here.
+	auto eLab = std::make_unique<ThreadFinishLabel>(pos, retVal);
+
+	GenMCDriver::handleThreadFinish(std::move(eLab));
+}
+
+/**** Memory access handling ****/
+
+[[nodiscard]] auto MiriGenMCShim::handleLoad(ThreadId thread_id, uint64_t address, uint64_t size,
+											 MemOrdering ord, GenmcScalar old_val) -> LoadResult
+{
+	auto pos = incPos(thread_id);
+
+	auto loc = SAddr(address);
+	auto aSize = ASize(size);
+	auto type = AType::Unsigned; // FIXME(genmc): get correct type from Miri(?)
+
+	auto newLab = std::make_unique<ReadLabel>(pos, ord, loc, aSize, type);
+
+	auto result = GenMCDriver::handleLoad(std::move(newLab));
+	return result;
+}
+
+[[nodiscard]] auto MiriGenMCShim::handleStore(ThreadId thread_id, uint64_t address, uint64_t size,
+											  GenmcScalar value, GenmcScalar old_val,
+											  MemOrdering ord)
+	-> StoreResult
+{
+	auto pos = incPos(thread_id);
+
+	auto loc = SAddr(address);
+	auto aSize = ASize(size);
+	auto type = AType::Unsigned; // FIXME(genmc): get correct type from Miri(?)
+
+	auto val = value.toSVal();
+
+	std::unique_ptr<WriteLabel> wLab = std::make_unique<WriteLabel>(pos, ord, loc, aSize, type, val);
+
+	return GenMCDriver::handleStore(std::move(wLab));
+}
+
+/**** Memory (de)allocation ****/
+
+auto MiriGenMCShim::handleMalloc(ThreadId thread_id, uint64_t size, uint64_t alignment) -> uintptr_t
+{
+	auto pos = incPos(thread_id);
+
+	// FIXME(genmc): get correct values from Miri
+	auto sd = StorageDuration::SD_Heap;
+	auto stype = StorageType::ST_Durable;
```

Comment:
Should be volatile.

Also add a comment clarifying that genmc ignores them except for printing (?)